### PR TITLE
Admission webhoooks

### PR DESCRIPTION
**Pull request by awgreene**
This PR introduces OLM support for validating and mutating admission
webhooks by updating the ClusterServiceVersion CRD to include a
WebhooksDefinitions array. In the current iteration, OLM supports
cycling the CA Certs required for webhooks.

### CODE REVIEW

File: pkg/api/wrappers/deployment_install_client.go [hunk 1]
```diff
@@ -45,6 +47,14 @@ func NewInstallStrategyDeploymentClient(opClient operatorclient.ClientInterface,
 	}
 }
 
+func (c *InstallStrategyDeploymentClientForNamespace) GetOpClient() operatorclient.ClientInterface {
```
Comment (awgreene) [inline pkg/api/wrappers/deployment_install_client.go:50]: Should be cleaned up in refactor.

File: pkg/api/wrappers/deployment_install_client.go [hunk 2]
```diff
@@ -45,6 +47,14 @@ func NewInstallStrategyDeploymentClient(opClient operatorclient.ClientInterface,
...
+	return c.opClient
+}
+
+func (c *InstallStrategyDeploymentClientForNamespace) GetOpLister() operatorlister.OperatorLister {
```
Comment (awgreene) [inline pkg/api/wrappers/deployment_install_client.go:54]: Should be cleaned up in refactor.

File: pkg/controller/operators/olm/operator.go [hunk 3]
```diff
@@ -1415,27 +1416,13 @@ func (a *Operator) transitionCSVState(in v1alpha1.ClusterServiceVersion) (out *v
...
-			if err != nil {
-				return
-			}
-
-			// Cleanup legacy resources
-			err = a.deleteLegacyAPIServiceResources(out, desc)
-			if err != nil {
-				return
-			}
+		if len(out.GetAllAPIServiceDescriptions()) > 0 {
```
Comment (awgreene) [inline pkg/controller/operators/olm/operator.go:1419]: This should be revisited.

File: pkg/controller/install/apiservice.go [hunk 4]
```diff
@@ -0,0 +1,803 @@
...
+}
+
+func (i *StrategyDeploymentInstaller) createOrUpdateWebhook(caPEM []byte, desc v1alpha1.WebhookDescription) error {
+	operatorGroups, err := i.strategyClient.GetOpLister().OperatorsV1().OperatorGroupLister().OperatorGroups(i.owner.GetNamespace()).List(labels.Everything())
+	if err != nil || len(operatorGroups) != 1 {
+		return fmt.Errorf("Error retrieving OperatorGroup info")
+	}
+	ogNamespacelabelSelector := operatorGroups[0].NamespaceLabelSelector()
+
+	switch desc.Type {
```
Comment (awgreene) [inline pkg/controller/install/apiservice.go:460]: Will need to figure out how to improve this function.

File: pkg/controller/install/apiservice.go [hunk 5]
```diff
@@ -0,0 +1,803 @@
...
+	"k8s.io/apimachinery/pkg/labels"
+	apiregistrationv1 "k8s.io/kube-aggregator/pkg/apis/apiregistration/v1"
+
+	"github.com/operator-framework/operator-lifecycle-manager/pkg/api/apis/operators/v1alpha1"
+	"github.com/operator-framework/operator-lifecycle-manager/pkg/controller/certs"
+	"github.com/operator-framework/operator-lifecycle-manager/pkg/lib/operatorlister"
+	"github.com/operator-framework/operator-lifecycle-manager/pkg/lib/ownerutil"
+)
+
+const (
```
Comment (awgreene) [inline pkg/controller/install/apiservice.go:25]: These constants exist in two files right now, should be refactored before merging.

File: pkg/api/apis/operators/clusterserviceversion_types.go [hunk 6]
```diff
@@ -127,6 +129,87 @@ type APIServiceDescription struct {
...
+				Path:      w.WebhookPath,
+			},
+			CABundle: caBundle,
+		},
+		ReinvocationPolicy: w.ReinvocationPolicy,
+	}
+}
+
+// MorphDomainName returns the result of replacing all periods in the given Webhook name with hyphens
+func (w *WebhookDescription) MorphDomainName() string {
```
Comment (ecordell) [inline pkg/api/apis/operators/clusterserviceversion_types.go:208]: nit: something more like `DomainName` or `ToDNSLabel`, etc

File: pkg/controller/install/webhook.go [hunk 7]
```diff
@@ -0,0 +1,43 @@
...
+		}
+
+		if contains(apiGroupMap, "operators.coreos.com") {
+			return fmt.Errorf("Webhook rules cannot include the OLM group")
+		}
+
+		// protect Admission Webhook resources
+		if contains(apiGroupMap, "admissionregistration.k8s.io") {
+			resourceGroupMap := listToMap(rule.Resources)
+			if contains(resourceGroupMap, "*") || contains(resourceGroupMap, "MutatingWebhookConfiguration") || contains(resourceGroupMap, "ValidatingWebhookConfiguration") {
```
Comment (estroz) [inline pkg/controller/install/webhook.go:25]: Might want to return an error specific to having `*` as a resource identifier: `Webhook rules cannot match all resources`.

File: pkg/controller/operators/olm/operator.go [hunk 8]
```diff
@@ -1415,27 +1416,13 @@ func (a *Operator) transitionCSVState(in v1alpha1.ClusterServiceVersion) (out *v
...
+			expiration := time.Now().Add(DefaultCertValidFor)
```
Comment (estroz) [inline pkg/controller/operators/olm/operator.go:1420]: For consistent start times:
```suggestion
			now := metav1.Now()
			expiration := now.DeepCopy().Add(DefaultCertValidFor)
```

File: pkg/api/apis/operators/clusterserviceversion_types.go (hunk 6 above)
Comment (awgreene) [reply pkg/api/apis/operators/clusterserviceversion_types.go]: Updated the function to `DomainName`.

File: pkg/controller/install/webhook.go (hunk 7 above)
Comment (awgreene) [reply pkg/controller/install/webhook.go:25]: To be clear, Webhooks are allowed to match all resources in an APIGroup but they can't intercept Webhook resources.

File: pkg/controller/operators/olm/operator.go (hunk 8 above)
Comment (awgreene) [reply pkg/controller/operators/olm/operator.go:1420]: Good idea!

Comment (awgreene) [reply pkg/controller/operators/olm/operator.go]: Addressed in the latest version of the PR.

**Comment by jmazzitelli** [comment]
How will this be exposed to Ansible operators?

**Comment by awgreene** [comment]
/retest

The following tests failed but ran locally
```
creation with dependencies [It]
image update [It]
```

**Comment by ecordell** [comment]
> How will this be exposed to Ansible operators?

That's probably a question for operator-sdk. This is for packaging / installing an admission webhook with your operator, and doesn't have any opinions about how the webhook is implemented.

File: pkg/api/apis/operators/clusterserviceversion_types.go [hunk 9]
```diff
@@ -163,6 +246,7 @@ type ClusterServiceVersionSpec struct {
 	Maturity                  string
 	CustomResourceDefinitions CustomResourceDefinitions
 	APIServiceDefinitions     APIServiceDefinitions
+	WebhookDefinitions        []WebhookDescription
```
Comment (harishsurf) [inline pkg/api/apis/operators/clusterserviceversion_types.go:249]: typo: the slice should be `[]WebhookDefinitions` ?

File: pkg/api/apis/operators/v1alpha1/clusterserviceversion.go [hunk 10]
```diff
@@ -133,6 +133,14 @@ func (c *ClusterServiceVersion) IsCopied() bool {
 	return false
 }
 
+func (c *ClusterServiceVersion) HasCAResources() bool {
+	// Return early if there are no owned APIServices
+	if len(c.Spec.APIServiceDefinitions.Owned)+len(c.Spec.WebhookDefinitions) == 0 {
```
Comment (harishsurf) [inline pkg/api/apis/operators/v1alpha1/clusterserviceversion.go:138]: nit: `return !(len(c.Spec.APIServiceDefinitions.Owned)+len(c.Spec.WebhookDefinitions) == 0)` ?

File: pkg/api/apis/operators/v1alpha1/clusterserviceversion_types.go [hunk 11]
```diff
@@ -191,6 +276,8 @@ type ClusterServiceVersionSpec struct {
 	CustomResourceDefinitions CustomResourceDefinitions `json:"customresourcedefinitions,omitempty"`
 	APIServiceDefinitions     APIServiceDefinitions     `json:"apiservicedefinitions,omitempty"`
 	// +listType=set
+	WebhookDefinitions []WebhookDescription `json:"webhookdefinitions,omitempty"`
```
Comment (harishsurf) [inline pkg/api/apis/operators/v1alpha1/clusterserviceversion_types.go:279]: typo: the slice should be []WebhookDefinitions ?

File: deploy/ocp/manifests/latest [hunk 12]
```diff
@@ -1 +1 @@
-./0.14.1
\ No newline at end of file
+./0.14.2
```
Comment (ecordell) [inline deploy/ocp/manifests/latest:1]: can you make sure to make the release in a separate commit?

(or, if you don't want to couple the release process to this PR, can you commit only /deploy/chart and /manifest changes and leave the others out?)

Comment (ecordell) [reply deploy/ocp/manifests/latest:1]: we're not overly strict about this (yet) but rule of thumb is that .z bumps are for small patches/fixes. A lot has changed since 0.14.1, so 0.15.0 seems reasonable.

File: pkg/controller/install/caresources.go [hunk 13]
```diff
@@ -0,0 +1,543 @@
+package install
```
Comment (ecordell) [inline pkg/controller/install/caresources.go:1]: naming nit: the filename also reads as `care sources` - what about something like "certs"?

File: pkg/controller/install/deployment.go [hunk 14]
```diff
@@ -59,13 +61,25 @@ func (c DeploymentInitializerFuncChain) Apply(deployment *appsv1.Deployment) (er
...
-		owner:               owner,
-		previousStrategy:    previousStrategy,
-		templateAnnotations: templateAnnotations,
-		initializers:        initializers,
+		strategyClient:         strategyClient,
+		owner:                  owner,
+		previousStrategy:       previousStrategy,
+		templateAnnotations:    templateAnnotations,
+		initializers:           initializers,
+		apiServiceDescriptions: apiDescs,
```
Comment (ecordell) [inline pkg/controller/install/deployment.go:81]: refactor suggestion: is there a way we can model the apidescriptions/webhookdescriptions as just more `initializer` functions?

File: pkg/controller/install/webhook.go [hunk 15]
```diff
@@ -0,0 +1,146 @@
...
+}
+
+func listToMap(list []string) map[string]struct{} {
+	result := make(map[string]struct{})
+	for _, ele := range list {
+		result[ele] = struct{}{}
+	}
+	return result
+}
+func contains(m map[string]struct{}, tar string) bool {
```
Comment (ecordell) [inline pkg/controller/install/webhook.go:47]: nit: spacing

File: pkg/controller/operators/olm/operator.go [hunk 16]
```diff
@@ -1366,6 +1367,14 @@ func (a *Operator) transitionCSVState(in v1alpha1.ClusterServiceVersion) (out *v
 			return
 		}
 
+		// Check if Webhooks have valid rules
+		for _, desc := range out.Spec.WebhookDefinitions {
+			if syncError = install.ValidWebhookRules(desc.Rules); syncError != nil {
```
Comment (ecordell) [inline pkg/controller/operators/olm/operator.go:1372]: future nit: we should tie these validations to the the validation library used in opm

File: pkg/controller/operators/olm/operator_test.go [hunk 17]
```diff
@@ -17,6 +16,8 @@ import (
 	"testing"
 	"time"
 
+	"github.com/operator-framework/operator-lifecycle-manager/pkg/lib/queueinformer"
```
Comment (ecordell) [inline pkg/controller/operators/olm/operator_test.go:19]: imports nit: we try to put the local imports in the last block

File: test/e2e/csv_e2e_test.go [hunk 18]
```diff
@@ -3315,219 +3316,1048 @@ var _ = Describe("CSV", func() {
 
 		require.Equal(GinkgoT(), apiService1.Spec.CABundle, apiService2.Spec.CABundle)
 	})
-})
+	It("creates global webhooks", func() {
```
Comment (ecordell) [inline test/e2e/csv_e2e_test.go:3319]: this can be a follow up, but we're trying to write new tests in a more BDD style, check out harish's PR: https://github.com/operator-framework/operator-lifecycle-manager/pull/1424

The tests look good otherwise, just a style thing.

File: pkg/api/apis/operators/clusterserviceversion_types.go (hunk 9 above)
Comment (awgreene) [reply pkg/api/apis/operators/clusterserviceversion_types.go:249]: I wanted to keep it similar to `APIServiceDefinitions`, I'm not opposed to changing the variable (API) name though.

File: pkg/api/apis/operators/v1alpha1/clusterserviceversion.go (hunk 10 above)
Comment (awgreene) [reply pkg/api/apis/operators/v1alpha1/clusterserviceversion.go:138]: I don't believe that this suggestion makes the code more readable.

File: pkg/api/apis/operators/v1alpha1/clusterserviceversion_types.go (hunk 11 above)
Comment (awgreene) [reply pkg/api/apis/operators/v1alpha1/clusterserviceversion_types.go:279]: See [this comment](https://github.com/operator-framework/operator-lifecycle-manager/pull/1436#discussion_r408147800)

File: test/e2e/csv_e2e_test.go (hunk 18 above)
Comment (awgreene) [reply test/e2e/csv_e2e_test.go:3319]: Noted.

File: deploy/ocp/manifests/latest (hunk 12 above)
Comment (awgreene) [reply deploy/ocp/manifests/latest]: In the interest of getting this in quickly I'll just commit the deploy/chart and /manifest changes for now.

File: pkg/controller/install/caresources.go (hunk 13 above)
Comment (awgreene) [reply pkg/controller/install/caresources.go]: Fair point - renamed to `certResources`.

**Comment by ecordell** [comment]
/approve

(to get the bots on this)

**Comment by awgreene** [comment]
Closed/Reopened to kick tests.

**Comment by awgreene** [comment]
/retest

CI Failures

**Comment by awgreene** [comment]
Closed and reopened to kick off tests after another PR was merged.

**Comment by awgreene** [comment]
Passing e2e-aws, e2e-gcp is broken right now - manually merging as a result
